subprojects/core/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheService.java

```diff
@@ -152,17 +167,11 @@ public void run() {
     }
 
     @Override
-    public void load(final BuildCacheKey key, final Action<? super File> reader) {
-        persistentCache.withFileLock(new Factory<Void>() {
+    public void allocateTempFile(final BuildCacheKey key, final Action<? super File> action) {
+        persistentCache.useCache(new Runnable() {
```

Comment:
Locking here slows things down unnecessarily. Instead we could make sure the cleanup only clears files that match `/[\da-f]{32}/i`.